File: typescript/src/server/rpc-protocol-plugin.ts
import type { Server as HttpServer } from 'node:http'
import { WebSocketServer, type WebSocket } from 'ws'
import { DEFAULT_RPC_PATH, type PLATRPCRequest } from '../rpc'
import { HttpError, type RouteContext } from '../types'
import type { PLATServerHostContext, PLATServerProtocolPlugin, PLATServerTransportRuntime } from './protocol-plugin'

export interface RpcProtocolPluginOptions {
  /** true, false, or a custom path like '/ws' */
  enabled: boolean | string
}

interface RpcProtocolPluginInternal extends PLATServerProtocolPlugin {
  /** Set by setup(), used by attachRpcProtocolPlugin */
  _runtime?: PLATServerTransportRuntime
}

export function createRpcProtocolPlugin(options: RpcProtocolPluginOptions): PLATServerProtocolPlugin {
  const plugin: RpcProtocolPluginInternal = {
    name: 'rpc',
    setup(rt: PLATServerTransportRuntime) {
      plugin._runtime = rt
    },
    attach(rt: PLATServerTransportRuntime, host: PLATServerHostContext) {
      attachRpcProtocolPlugin(plugin, options, host)
    },
    teardown() {
      plugin._runtime = undefined
    },
  }
  return plugin
}

/**
 * Attach the RPC WebSocket server to an existing HTTP server.
 * Must be called after plugin.setup(runtime) and before server.listen().
 */
export function attachRpcProtocolPlugin(
  plugin: PLATServerProtocolPlugin,
  options: RpcProtocolPluginOptions,
  host: PLATServerHostContext,
): void {
  if (options.enabled === false || plugin.name !== 'rpc') return

  const rt = (plugin as RpcProtocolPluginInternal)._runtime
  if (!rt) return
  if (host.kind !== 'node-http' || !host.server) return

  const rpcPath = typeof options.enabled === 'string' ? options.enabled : DEFAULT_RPC_PATH
  const serialize = (v: unknown) => rt.serializeValue(v)
  const activeRpcCalls = new Map<string, AbortController>()

  const server = host.server as HttpServer
  const wss = new WebSocketServer({ server, path: rpcPath })

  // Store for teardown
  const originalTeardown = plugin.teardown
  plugin.teardown = async (runtime) => {
    for (const ac of activeRpcCalls.values()) ac.abort()
    activeRpcCalls.clear()
    wss.close()
    await originalTeardown?.(runtime)
  }

  wss.on('connection', (socket: WebSocket) => {
    socket.on('message', async (message: Parameters<WebSocket['send']>[0]) => {
      const raw = typeof message === 'string' ? message : message.toString()

      let request: PLATRPCRequest
      try {
        request = JSON.parse(raw) as PLATRPCRequest
      } catch {
        socket.send(JSON.stringify({
          jsonrpc: '2.0', id: 'invalid', ok: false,
          error: { status: 400, message: 'Invalid RPC payload' },
        }))
        return
      }

      if (request.cancel) {
        activeRpcCalls.get(request.id)?.abort()
        socket.send(JSON.stringify({
          jsonrpc: '2.0', id: request.id, ok: true, result: { cancelled: true },
        }))
        return
      }

      const operation = rt.resolveOperation({
        operationId: request.operationId,
        method: request.method,
        path: request.path,
      })

      if (!operation) {
        socket.send(JSON.stringify({
          jsonrpc: '2.0', id: request.id, ok: false,
          error: { status: 404, message: `RPC operation not found for ${request.method} ${request.path}` },
        }))
        return
      }

      const abortController = new AbortController()
      activeRpcCalls.set(request.id, abortController)

      try {
        const normalizedInput = rt.normalizeInput(
          typeof request.input === 'object' && request.input !== null ? request.input as Record<string, any> : {},
        )

        const ctx: RouteContext = {
          method: operation.method,
          url: operation.path,
          headers: request.headers ?? {},
          opts: operation.routeMeta?.opts,
        }

        rt.createCallContext({
          ctx,
          sessionId: request.id,
          mode: 'rpc',
          signal: abortController.signal,
          emit: async (event: 'progress' | 'log' | 'chunk' | 'message', data?: unknown) => {
            socket.send(JSON.stringify({
              jsonrpc: '2.0', id: request.id, ok: true, event,
              data: serialize(data),
            }))
          },
        })

        const envelope = rt.createEnvelope({
          protocol: 'rpc',
          operation,
          input: normalizedInput,
          headers: request.headers ?? {},
          ctx,
          requestId: request.id,
          req: { headers: request.headers ?? {} },
          allowHelp: false,
          helpRequested: false,
        })

        const execution = await rt.dispatch(operation, envelope)
        socket.send(JSON.stringify({
          jsonrpc: '2.0', id: request.id, ok: true,
          result: serialize(execution.result),
        }))
      } catch (error: any) {
        const status = error instanceof HttpError ? error.statusCode : 500
        socket.send(JSON.stringify({
          jsonrpc: '2.0', id: request.id, ok: false,
          error: {
            status,
            message: error?.message ?? 'Internal server error',
            data: error instanceof HttpError ? error.data : undefined,
          },
        }))
      } finally {
        activeRpcCalls.delete(request.id)
      }
    })
  })
}

File: python/plat/rpc_protocol_plugin.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .rpc import DEFAULT_RPC_PATH
from .protocol_plugin import ServerHostContext, ServerTransportRuntime


@dataclass
class RpcProtocolPluginOptions:
    """Configuration for the RPC protocol plugin."""
    enabled: bool | str = True
    http_error_class: type | None = None  # HttpError class for status code extraction


def create_rpc_protocol_plugin(options: RpcProtocolPluginOptions) -> Any:
    """Create a self-contained RPC (WebSocket) protocol plugin for the Python server."""

    if options.enabled is False:
        class _NoOpPlugin:
            name = "rpc"
            def setup(self, runtime: Any) -> None: pass
            def attach(self, runtime: Any, host: Any) -> None: pass
            def start(self, runtime: Any) -> None: pass
        return _NoOpPlugin()

    rpc_path = options.enabled if isinstance(options.enabled, str) else DEFAULT_RPC_PATH
    HttpError = options.http_error_class

    runtime_ref: list[ServerTransportRuntime | None] = [None]
    app_ref: list[Any | None] = [None]

    async def send_event(websocket: Any, request_id: str, event: str, data: Any = None) -> None:
        await websocket.send_text(json.dumps({
            "jsonrpc": "2.0",
            "id": request_id,
            "ok": True,
            "event": event,
            "data": serialize(data),
        }))

    async def handle_payload(payload: dict[str, Any], websocket: Any) -> dict[str, Any]:
        rt = runtime_ref[0]
        if rt is None:
            return {"jsonrpc": "2.0", "id": "error", "ok": False, "error": {"status": 500, "message": "RPC runtime not ready"}}

        request_id = str(payload.get("id", "invalid"))

        if payload.get("cancel"):
            # Cancel not yet wired (no abort controller in Python), but ack it
            return {"jsonrpc": "2.0", "id": request_id, "ok": True, "result": {"cancelled": True}}

        operation = rt.resolve_operation(
            operation_id=payload.get("operationId") if isinstance(payload.get("operationId"), str) else None,
            method=str(payload.get("method", "")).upper(),
            path=str(payload.get("path", "")),
        )
        if operation is None:
            return {
                "jsonrpc": "2.0", "id": request_id, "ok": False,
                "error": {"status": 404, "message": "RPC operation not found"},
            }

        try:
            input_data = payload.get("input") or {}
            if not isinstance(input_data, dict):
                input_data = {}
            input_data = rt.normalize_input(input_data)

            headers = payload.get("headers") or {}
            route_meta = operation.route_meta

            from .server_types import RouteContext
            ctx = RouteContext(
                method=operation.method,
                url=operation.path,
                headers=dict(headers),
                opts=route_meta.opts if route_meta else None,
            )

            async def emit(event: str, data: Any = None) -> None:
                await send_event(websocket, request_id, event, data)

            rt.create_call_context(
                ctx=ctx,
                session_id=request_id,
                mode="rpc",
                emit=emit,
                signal=None,
            )

            execution = await rt.dispatch(
                operation,
                rt.create_envelope(
                    protocol="rpc",
                    operation=operation,
                    input=input_data,
                    ctx=ctx,
                    headers=dict(headers),
                    request_id=request_id,
                    request=type("RpcRequest", (), {"headers": dict(headers)})(),
                    allow_help=False,
                    help_requested=False,
                ),
            )

            result = execution.get("result") if isinstance(execution, dict) else execution
            # Handle http_response execution kind
            if isinstance(execution, dict) and execution.get("kind") == "http_response":
                result = rt.serialize_value(getattr(execution.get("http_response"), "body", result))
            else:
                result = rt.serialize_value(result)

            return {"jsonrpc": "2.0", "id": request_id, "ok": True, "result": result}

        except Exception as exc:
            status = 500
            if HttpError and isinstance(exc, HttpError):
                status = exc.status_code
            else:
                status = getattr(exc, "status_code", None) or getattr(exc, "status", None) or 500
            return {
                "jsonrpc": "2.0", "id": request_id, "ok": False,
                "error": {
                    "status": status,
                    "message": str(exc) or "Internal server error",
                    **({"data": getattr(exc, "data")} if getattr(exc, "data", None) is not None else {}),
                },
            }

    class RpcProtocolPlugin:
        name = "rpc"

        def setup(self, runtime: ServerTransportRuntime) -> None:
            runtime_ref[0] = runtime

        def attach(self, runtime: ServerTransportRuntime, host: ServerHostContext) -> None:
            if host.kind != "fastapi-uvicorn":
                return
            app_ref[0] = host.app

        def start(self, runtime: ServerTransportRuntime) -> None:
            # Register the WebSocket route on the FastAPI app
            app = app_ref[0]
            if app is None:
                return

            try:
                from fastapi import WebSocket as WSType
            except ImportError:
                return

            @app.websocket(rpc_path)
            async def rpc_socket(websocket: WSType):
                await websocket.accept()
                while True:
                    try:
                        payload = await websocket.receive_json()
                    except Exception:
                        break
                    response = await handle_payload(payload, websocket)
                    await websocket.send_text(json.dumps(response))

        def teardown(self, runtime: ServerTransportRuntime) -> None:
            runtime_ref[0] = None

    return RpcProtocolPlugin()
